### PR TITLE
added ClimWeb for 17 African countries

### PR DESCRIPTION
**Pull request by chunshek**
I have added ClimWeb, an open source CMS developed by WMO Africa, for 17 African countries. I have implemented the following:

- Alerts
- Temperature normals

### CODE REVIEW

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        }
+    }
+
+    // Location parameters
+    override fun needsLocationParametersRefresh(
+        location: Location,
+        coordinatesChanged: Boolean,
+        features: List<SourceFeature>,
+    ): Boolean {
+        // As a secondary source, this function does not receive any value for features.
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:175]: Fixed in 3789ad4c2d452ac857825a001c3d4d038653612e

Comment (chunshek) [reply app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:175]: Oh nice! I will incorporate that change.

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            )
+        }
+    }
+
+    // CONFIG
+    private val config
+        get() = SourceConfigStore(context, id)
+    override val isConfigured = true
+    override val isRestricted = false
+    private var instance: String
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:227]: There were recent changes in Preferences to allow to reset the instance to default URL. Change to `String?`

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        set(value) {
+            config.edit().putString("instance", value).apply()
+        }
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:230]:
```kotlin
set(value) {
            value?.let {
                config.edit().putString("instance", value).apply()
            } ?: config.edit().remove("instance").apply()
}
```

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    override fun getPreferences(context: Context): List<Preference> {
+        return listOf(
+            EditTextPreference(
+                titleId = instancePreference,
+                summary = { _, content ->
+                    content.ifEmpty {
+                        baseUrl
+                    }
+                },
+                content = instance,
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:242]: `content = if (instance != baseUrl) instance else null,`

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                regex = EditTextPreference.URL_REGEX,
+                regexError = context.getString(R.string.settings_source_instance_invalid),
+                onValueChanged = {
+                    instance = it
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:246]: `instance = if (it == baseUrl) null else it.ifEmpty { null }`

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+     */
+    protected abstract val instancePreference: Int
+
+    // TODO: Change this to override when ClimWeb is ready to be used as MainWeatherSource
+    protected abstract val weatherAttribution: String
+
+    protected val mApi: ClimWebApi
+        get() {
+            return jsonClient
+                .baseUrl(instance)
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:90]: `.baseUrl(instance!!)` (see below why)

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt (hunk 2 above)
Comment (chunshek) [reply app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:227]: Changed.

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt (hunk 3 above)
Comment (chunshek) [reply app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:230]: Changed.

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt (hunk 4 above)
Comment (chunshek) [reply app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:242]: Changed and added.

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt (hunk 5 above)
Comment (chunshek) [reply app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:246]: Changed.

File: app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt (hunk 6 above)
Comment (chunshek) [reply app/src/main/java/org/breezyweather/sources/climweb/ClimWebService.kt:90]: Changed.

**Comment by chunshek** [comment]
I'll compile with all the changes above and test it before pushing it again.

**Pull request merged**